### ISSUE TITLE
is mapcontroller disposed after setState ?

### ISSUE THREAD

**Issue by an99990** [fleaflet/flutter_map#1114]
I am new to flutter and trying to build a map with different markers depending on the zoom. Here is the code. Basically when we finish zooming, i check if the value is within a certain range to update the markers.

```
StreamSubscription subscription;
bool _circleSigns = false;

 @override
  void initState() {
    super.initState();

    widget.controller.mapEventStream.listen((data) {
      _updateMarkers(data);
    });
  }

   void _updateMarkers(mapEvent) {
widget.controller.onReady.then((_) {
  subscription =
      widget.controller.mapEventStream.listen((MapEvent mapEvent) {
    // is we finished zooming and our
    if (mapEvent is MapEventMoveEnd) {        
      // if we are far, we have to change the marker to parkings only if circles are already there
      if (mapEvent.zoom >= _maxZoom && _circleSigns) {
        setState(() {
          updatedMarkers = drawingMap.Markers();
          // we just changed the map to circle markers no need to change it after
          _circleSigns = false;
         // widget.controller = new MapController();
        });
      }
      // if we are close, we change the markers to circle, only if they arent already there
      if (mapEvent.zoom < _maxZoom && !_circleSigns) {
        setState(() {
          updatedMarkers = drawingMap.CircleMarkers();
          // we just switched to parking signs
          _circleSigns = true;
         // widget.controller = new MapController();
        });
      }
    }
  });
});
```
However, when i call setState, the app crashes and i get
`
    ════════ Exception caught by gesture ═══════════════════════════════════════════ Bad state: Cannot add new events after calling close`

I found a workaround where I would create a new mapController after i setstate, but now the controller doesnt seem to be ever ready..

```
 return Scaffold(
      body: buildMap(data,  markers),
      floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.navigation),
          onPressed: () {
            widget.controller.onReady.then((_) {
              widget.controller.move(
                  latLng.LatLng(data.latitude, data.longitude),
                  17.0);
            });
          }),
      drawer: DrawerMenu(),
    );
```
Not exactly sure what the problem is, i understand that creating a new controller at every state isnt the best solution, but if i dont do it, i always get the error mentioned earlier...

 Here is my initstate
```

  void initState() {
    super.initState();
    location = new Location();
   widget.controller = new MapController();
    subscription = widget.controller.mapEventStream.listen((data) {
      _updateMarkers(data);
    });
    // Timer.periodic(Duration(seconds: 30), (_timer) => );
  }
```
please help

**Comment by ibrierley** [comment]
It's not really clear from the code which setState is being called, for which widget as the rest of the code isn't included.

It feels a bit weird to have a repeated onready assignment within an _updateMarkers type code as well, but I may be wrong, but it all on first glance looks a bit messy and liable to bugs like that. It looks like when you get a mapController event, you then call updateMarkers which then sets up a new onReady event to listen to a stream which is already listened to (or something, it's hard to follow, which feels like this is part of the problem in itself).

I think you need to edit the code to include the full (or minimal reproducable widget)

**Comment by an99990** [comment]
Thank you for the answer, here is the minimal code. Sorry if it is messy, this is my first time. Basically i return different map depending on the zoom level. I think since i am creating a new Controller at each setState, i have to tell it again to listen to the stream, otherwise i dont get the marker that i want.. 

```
class MainMapBody extends StatefulWidget {
  MapController controller = new MapController();
  bool serviceEnabled;
  PermissionStatus permissionGranted;

  @override
  _MainMapBodyState createState() => _MainMapBodyState();
}

class _MainMapBodyState extends State<MainMapBody> {
  Location location;
  LocationData locationData;
  var currentLocation; // hold the currentlocation of the user

  var updatedMarkers; // Depending on the zoom updatedMarkers can be List<Marker> or List<CircleMarker>
  var _maxZoom; // maxZoom value to reach before changing the marker
  var _apiTaxiData; // Holds all the information retrieved from network/cache
  bool _circleSigns = false; // holds the current merker on device
  var _drawingMap; // holds all the markers

  var mSnapshot; // Holds information returned by futurebuilder
  var firstState = true; // Allow to know if we built the map already
  var permissionGranted; // holds the permission status
  StreamSubscription<MapEvent> subscription; // hold the zooming event

  _MainMapBodyState() {
    _maxZoom = 13.0;
  }

  @override
  void initState() {
    super.initState();
    widget.controller = new MapController();
    location = new Location();
    widget.controller.mapEventStream
        .where((event) => event is MapEventMoveEnd)
        .listen((data) {
      _updateMarkers(data);
    });
  }

  void _updateMarkers(mapEvent) {
// here since i create another controller at setstate, i need to tell it to relisten to the stream ...
    widget.controller.onReady.then((_) {
      // only listen when mapEventEnd
      subscription = widget.controller.mapEventStream
          .where((event) => event is MapEventMoveEnd)
          .listen((MapEvent mapEvent) {
        firstState = false;
        print(mapEvent.zoom.toString());
        // if we are far, we have to change the marker to parkings only if circles are already there
        if (mapEvent.zoom >= _maxZoom && _circleSigns) {
          widget.controller = new MapController();
          updateMarkers();
        }
        // if we are close, we change the markers to circle, only if they arent already there
        if (mapEvent.zoom < _maxZoom && !_circleSigns) {
          widget.controller = new MapController();
          updateCircleMarkers();
        }
      });
    });
  }

  void updateMarkers() {
    updatedMarkers = _drawingMap.getMarkers();
    // we just changed the map to circle markers no need to change it after
    _circleSigns = false;
    setState(() {});
    // widget.controller = new MapController();
  }

  void updateCircleMarkers() {
    updatedMarkers = _drawingMap.getCircleMarkers();
    _circleSigns = true;
    setState(() {});
    // widget.controller = new MapController();
  }

  @override
  Widget build(BuildContext context) {
    if (firstState) {
      return futureMap();
    } else {
      return updateMap();
    }
  }

  FutureBuilder futureMap() {
    return FutureBuilder(
        future: _getAsyncData(),
        builder: (context, snapshot) {
          mSnapshot = snapshot;
          if (snapshot.connectionState == ConnectionState.waiting) {
            return Container(
                child: Center(
                    child: SizedBox(
              child: CircularProgressIndicator(),
              height: 100.0,
              width: 100.0,
            )));
          } else if (snapshot.connectionState == ConnectionState.done) {
            if (snapshot.hasError) {
              return const Text('Error');
            } else if (snapshot.hasData) {
              if (updatedMarkers == null) {
                updatedMarkers = _drawingMap.getMarkers();
              }
              return updateMap();
            } else {
              return const Text('Empty data');
            }
          } else {
            return Text('State: ${snapshot.connectionState}');
          }
        });
  }

  Scaffold updateMap() {
    _getLocation().then((value) => mSnapshot = value);
    var currentData;
    if (mSnapshot is! latLng.LatLng) {
      currentData = mSnapshot.data;
    } else {
      currentData = mSnapshot;
    }

    return Scaffold(
//build map return the appropriate map with the desired marker
      body: buildMap(currentData, widget, updatedMarkers),
      floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.navigation),
          onPressed: () {
            widget.controller.onReady.then((_) {
              widget.controller.move(
                  latLng.LatLng(currentData.latitude, currentData.longitude),
                  17.0);
            });
          }),
    );
  }
}
```

let me know if you need more info, thank you for your help :)

**Comment by ibrierley** [comment]
I think if you are changing maps, it will likely be related to the problem. See if this post helps some idea...

https://github.com/fleaflet/flutter_map/issues/926 read the post near the end from @magnuswikhog

**Comment by an99990** [comment]
Thank you, i will give it a try !
